src/components/pages/devices/flyouts/deviceJobs/deviceJobTags.js

```diff
@@ -1,20 +1,304 @@
+    ];
+
+    // Link these values in render because they need to update based on component state
+    const tagLinks = this.tagsLink.getLinkedChildren(tagLink => {
+      const name = tagLink.forkTo('name')
+        .check(Validator.notEmpty, () => this.props.t('devices.flyouts.jobs.validation.required'));
+      const value = tagLink.forkTo('value')
+        .check(Validator.notEmpty, () => this.props.t('devices.flyouts.jobs.validation.required'));
+      const type = tagLink.forkTo('type')
+        .check(Validator.notEmpty, () => this.props.t('devices.flyouts.jobs.validation.required'));
+      const edited = !(!name.value && !value.value && !type.value);
+      const error = (edited && (name.error || value.error || type.error)) || '';
+      return { name, value, type, edited, error };
+    });
+    const editedTags = tagLinks.filter(({ edited }) => edited);
+    const tagsHaveErrors = editedTags.some(({ error }) => !!error);
+
+    return (
+      <form onSubmit={this.apply}>
+        <FormSection className="device-job-tags-container">
+          <SectionHeader>{t('devices.flyouts.jobs.tags.title')}</SectionHeader>
+          <SectionDesc>{t('devices.flyouts.jobs.tags.description')}</SectionDesc>
+
+          <FormGroup>
+            <FormLabel>{t('devices.flyouts.jobs.jobName')}</FormLabel>
+            <div className="help-message">{t('devices.flyouts.jobs.jobNameHelpMessage')}</div>
+            <FormControl className="long" link={this.jobNameLink} type="text" placeholder={t('devices.flyouts.jobs.jobNameHint')} />
+          </FormGroup>
+
+          <Grid className="data-grid">
+            <GridHeader>
+              <Row>
+                <Cell className="col-3">{t('devices.flyouts.jobs.tags.keyHeader')}</Cell>
+                <Cell className="col-3">{t('devices.flyouts.jobs.tags.valueHeader')}</Cell>
+                <Cell className="col-3">{t('devices.flyouts.jobs.tags.typeHeader')}</Cell>
+                <Cell className="col-1"></Cell>
+              </Row>
+              <Row className="action-row">
+                <Btn svg={svgs.plus} onClick={this.addTag}>{t('devices.flyouts.jobs.tags.add')}</Btn>
+              </Row>
+            </GridHeader>
+            <GridBody>
+              {
+                Object.keys(commonTags).length === 0 &&
+                <ErrorMsg className="device-jobs-error">{t('devices.flyouts.jobs.tags.noneExist')}</ErrorMsg>
+              }
+              {
+                Object.keys(commonTags).length > 0 &&
+                tagLinks.map(({ name, value, type, edited, error }, idx) => [
+                  <Row key={idx} className={error ? "error-data-row" : ""}>
+                    <Cell className="col-3">
+                      <FormControl className="shorter" type="text" link={name} errorState={!!error} />
+                    </Cell>
+                    <Cell className="col-3">
+                      <FormControl className="shorter" type="text" link={value} errorState={!!error} /></Cell>
```

Comment:
Haha, `shorter`. I'm not going to block the PR but put a TODO in the formControl to rename these. We should probably go with something like extra-small, small, regular .. etc. Better yet set it up to use 100% width correctly.